### PR TITLE
solution

### PR DESCRIPTION
**Pull request by victorgri**

### CODE REVIEW

**Comment by victorgri** [comment]
https://victorgri.github.io/react_dynamic-list-of-todos/

File: src/App.tsx [hunk 1]
```diff
@@ -1,14 +1,67 @@
...
 
+const debounce = (f: (query: string) => void, delay: number) => {
+  let timerId: number;
+
+  return (...args: string[]) => {
+    clearTimeout(timerId);
+
+    timerId = window.setTimeout(f, delay, ...args);
+  };
+};
```
Comment (SanyaBratashchuk) [inline src/App.tsx:21]: move it to the utils folder

File: src/App.tsx [hunk 2]
```diff
@@ -1,14 +1,67 @@
...
+
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[] | null>(null);
+  const [selectedTodo, setSelectedTodo] = useState<Todo | null>(null);
+  const [filter, setFilter] = useState('all');
```
Comment (SanyaBratashchuk) [inline src/App.tsx:26]: create enum for filter options

File: src/App.tsx [hunk 3]
```diff
@@ -1,14 +1,67 @@
...
+  const [query, setQuery] = useState('');
+  const [appliedQuery, setAppliedQuery] = useState('');
+
+  useEffect(() => {
+    getTodos()
+      .then(todosList => setTodos(todosList));
```
Comment (SanyaBratashchuk) [inline src/App.tsx:32]:
```suggestion
      .then(setTodos);
```

File: src/App.tsx [hunk 4]
```diff
@@ -1,14 +1,67 @@
...
+  }, [todos]);
```
Comment (SanyaBratashchuk) [inline src/App.tsx:33]: why do you need todos in an array of dependencies?

File: src/App.tsx [hunk 5]
```diff
@@ -1,14 +1,67 @@
...
+            && todo.title.toLowerCase().includes(filterQuery.toLowerCase());
+
+        case 'completed':
+          return todo.completed
+            && todo.title.toLowerCase().includes(filterQuery.toLowerCase());
+
+        default:
+          return todo;
+      }
+    });
```
Comment (SanyaBratashchuk) [inline src/App.tsx:60]: if it is a clear function you can move it out from the component

File: src/App.tsx [hunk 6]
```diff
@@ -17,18 +70,36 @@ export const App: React.FC = () => {
...
-              <TodoList />
+              {!todos && <Loader />}
+
+              {filteredTodos && (
+                <TodoList
+                  todos={filteredTodos}
+                  selectedTodo={selectedTodo}
+                  setSelectedTodo={setSelectedTodo}
+                />
+              )}
```
Comment (SanyaBratashchuk) [inline src/App.tsx:91]: You can simplify it `filteringTodos ? todoList : Loader`

File: src/App.tsx [hunk 7]
```diff
@@ -17,18 +70,36 @@ export const App: React.FC = () => {
...
             </div>
           </div>
         </div>
       </div>
 
-      <TodoModal />
+      {selectedTodo && (
+        <TodoModal
+          selectedTodo={selectedTodo}
+          setSelectedTodo={() => setSelectedTodo(null)}
```
Comment (SanyaBratashchuk) [inline src/App.tsx:100]:
```suggestion
          onClose={() => setSelectedTodo(null)}
```

File: src/components/TodoFilter/TodoFilter.tsx [hunk 8]
```diff
@@ -1,34 +1,67 @@
...
-        placeholder="Search..."
-      />
-      <span className="icon is-left">
-        <i className="fas fa-magnifying-glass" />
-      </span>
+type Props = {
+  setFilter: (filter: string) => void,
+  filter: string,
+  query: string,
+  setQuery: (query: string) => void,
```
Comment (SanyaBratashchuk) [inline src/components/TodoFilter/TodoFilter.tsx:7]:
```suggestion
  setQuery: React.Dispatch<React.SetStateAction<string>>,
```

File: src/components/TodoFilter/TodoFilter.tsx [hunk 9]
```diff
@@ -1,34 +1,67 @@
...
+        <input
+          data-cy="searchInput"
+          type="text"
+          className="input"
+          placeholder="Search..."
+          value={query}
+          onChange={({ target }) => {
+            setQuery(target.value);
+            applyQuery(target.value);
+          }}
```
Comment (SanyaBratashchuk) [inline src/components/TodoFilter/TodoFilter.tsx:44]: move it to the separate function

File: src/components/TodoFilter/TodoFilter.tsx [hunk 10]
```diff
@@ -1,34 +1,67 @@
...
+          <span className="icon is-right" style={{ pointerEvents: 'all' }}>
+            {/* eslint-disable-next-line jsx-a11y/control-has-associated-label */}
+            <button
+              data-cy="clearSearchButton"
+              type="button"
+              className="delete"
+              onClick={() => {
+                setQuery('');
+                applyQuery('');
+              }}
```
Comment (SanyaBratashchuk) [inline src/components/TodoFilter/TodoFilter.tsx:60]: move it to the separate function

File: src/components/TodoList/TodoList.tsx [hunk 11]
```diff
@@ -1,100 +1,80 @@
...
-          </span>
-        </th>
-        <th>Title</th>
-        <th> </th>
-      </tr>
-    </thead>
+type Props = {
+  todos: Todo[],
+  selectedTodo: Todo | null,
+  setSelectedTodo: (todo: Todo) => void,
```
Comment (SanyaBratashchuk) [inline src/components/TodoList/TodoList.tsx:7]: the same use more specific types

File: src/components/TodoModal/TodoModal.tsx [hunk 12]
```diff
@@ -1,12 +1,29 @@
...
+
+export const TodoModal: React.FC<Props> = ({
+  selectedTodo,
+  setSelectedTodo,
+}) => {
+  const [selectedUser, setSelectedUser] = useState<User | null>(null);
+
+  useEffect(() => {
+    getUser(selectedTodo.userId)
+      .then(user => setSelectedUser(user));
```
Comment (SanyaBratashchuk) [inline src/components/TodoModal/TodoModal.tsx:20]:
```suggestion
      .then(setSelectedUser);
```

**Review by SanyaBratashchuk — CHANGES_REQUESTED** [review]

**Pull request closed without merging**